File: native/Sources/Views/Timeline/TimelineLayout.swift
import Foundation

/// How much time the track shows at once.
enum TimelineZoom: String, CaseIterable, Identifiable, Sendable {
    case week, month, quarter, year

    var id: String { rawValue }

    var title: String {
        switch self {
        case .week: "周"
        case .month: "月"
        case .quarter: "季"
        case .year: "年"
        }
    }

    var span: TimeInterval {
        switch self {
        case .week: 7 * 86_400
        case .month: 30 * 86_400
        case .quarter: 91 * 86_400
        case .year: 365 * 86_400
        }
    }

    /// Unit and stride of the ruler's ticks, and the coarser unit whose boundaries are emphasised.
    fileprivate var ruler: (unit: Calendar.Component, step: Int, major: Calendar.Component) {
        switch self {
        case .week: (.day, 1, .weekOfYear)
        case .month: (.day, 3, .month)
        case .quarter: (.weekOfYear, 1, .month)
        case .year: (.month, 1, .quarter)
        }
    }
}

/// What the rows are grouped by. Both dimensions stay visible either way: grouping by directory
/// puts the agent's mark on every row, and grouping by agent puts the directory there instead.
enum TimelineGrouping: String, CaseIterable, Identifiable, Sendable {
    case directory, agent

    var id: String { rawValue }

    var title: String {
        switch self {
        case .directory: "按目录"
        case .agent: "按 Agent"
        }
    }
}

/// The visible span of time, in absolute dates. Panning and zooming are expressed here rather than
/// in the view so both are ordinary arithmetic that can be checked without a window on screen.
struct TimelineWindow: Equatable, Sendable {
    var start: Date
    var end: Date

    /// A window ending shortly after `anchor`, so "now" sits near the right edge with a little room
    /// to breathe rather than exactly on it.
    init(zoom: TimelineZoom, anchor: Date) {
        let span = zoom.span
        end = anchor.addingTimeInterval(span * 0.08)
        start = end.addingTimeInterval(-span)
    }

    init(start: Date, end: Date) {
        self.start = start
        self.end = min(start.addingTimeInterval(-1), end) == end ? start.addingTimeInterval(1) : end
    }

    var span: TimeInterval { max(1, end.timeIntervalSince(start)) }

    func fraction(of date: Date) -> Double {
        date.timeIntervalSince(start) / span
    }

    func x(of date: Date, width: CGFloat) -> CGFloat {
        CGFloat(fraction(of: date)) * width
    }

    func contains(_ date: Date) -> Bool {
        date >= start && date <= end
    }

    /// Moves by a share of the visible span. Positive moves forward in time.
    func shifted(byFractionOfSpan fraction: Double) -> TimelineWindow {
        let delta = span * fraction
        return TimelineWindow(
            start: start.addingTimeInterval(delta),
            end: end.addingTimeInterval(delta)
        )
    }

    /// Changes span while holding the right edge, which is where the eye already is: zooming out
    /// from today should reveal more history, not scroll away from the present.
    func zoomed(to zoom: TimelineZoom) -> TimelineWindow {
        TimelineWindow(start: end.addingTimeInterval(-zoom.span), end: end)
    }
}

/// A bar's horizontal extent inside a track of a given width.
struct TimelineBar: Equatable, Sendable {
    var x: CGFloat
    var width: CGFloat

    var maxX: CGFloat { x + width }
}

/// One labelled position on the ruler.
struct TimelineTick: Equatable, Sendable {
    var date: Date
    var label: String
    var major: Bool
}

/// A row: one session, drawn as one bar.
struct TimelineEntry: Identifiable, Equatable, Sendable {
    var session: HistorySessionMetadata

    var id: String { session.id }
    var start: Date { session.createdAt }
    /// A session that wrote a single record has identical timestamps; the layout gives it a minimum
    /// width rather than a zero-width bar nobody can click.
    var end: Date { max(session.lastActivity, session.createdAt) }
}

/// One horizontal lane inside a group: the other half of the pair the grouping did not take. A
/// directory group therefore has one lane per agent that worked in it, and an agent group one lane
/// per directory it touched — which is the whole question this view exists to answer, on one screen
/// instead of one row per session.
struct TimelineLane: Identifiable, Equatable, Sendable {
    var id: String
    var title: String
    /// Set when the lane *is* an agent, so its label can carry the brand mark.
    var source: HistorySource?
    var entries: [TimelineEntry]

    var lastActivity: Date { entries.map(\.end).max() ?? .distantPast }
}

struct TimelineGroup: Identifiable, Equatable, Sendable {
    var id: String
    var title: String
    var subtitle: String?
    /// Set when the group *is* an agent.
    var source: HistorySource?
    var lanes: [TimelineLane]

    var sessionCount: Int { lanes.reduce(0) { $0 + $1.entries.count } }
    var lastActivity: Date { lanes.map(\.lastActivity).max() ?? .distantPast }
}

enum TimelineLayout {
    /// Smallest bar that still reads as a bar and can be clicked.
    static let minimumBarWidth: CGFloat = 6

    static func bar(
        from start: Date,
        to end: Date,
        in window: TimelineWindow,
        width: CGFloat,
        minimum: CGFloat = minimumBarWidth
    ) -> TimelineBar? {
        guard width > 0, end >= start else { return nil }
        guard end >= window.start, start <= window.end else { return nil }
        let rawStart = window.x(of: start, width: width)
        let rawEnd = window.x(of: end, width: width)
        var x = max(0, rawStart)
        var barWidth = min(width, rawEnd) - x
        if barWidth < minimum {
            barWidth = minimum
            // Keep a clipped bar inside the track instead of letting the minimum push it out.
            x = min(x, width - minimum)
            x = max(0, x)
        }
        return TimelineBar(x: x, width: min(barWidth, width - x))
    }

    /// Vertical extent of a capsule inside a lane row, so the drawing and the hit test agree.
    static let capsuleTop: CGFloat = 7
    static let capsuleHeight: CGFloat = 16

    /// The session a click at `point` landed on, searched newest-drawn first so the capsule you can
    /// see on top of an overlap is the one that opens.
    static func entry(
        at point: CGPoint,
        in entries: [TimelineEntry],
        window: TimelineWindow,
        width: CGFloat,
        tolerance: CGFloat = 2
    ) -> TimelineEntry? {
        guard point.y >= capsuleTop - tolerance,
              point.y <= capsuleTop + capsuleHeight + tolerance else { return nil }
        for entry in entries.reversed() {
            guard let bar = bar(from: entry.start, to: entry.end, in: window, width: width)
            else { continue }
            if point.x >= bar.x - tolerance, point.x <= bar.maxX + tolerance { return entry }
        }
        return nil
    }

    static func ticks(
        in window: TimelineWindow,
        zoom: TimelineZoom,
        calendar: Calendar,
        locale: Locale
    ) -> [TimelineTick] {
        let ruler = zoom.ruler
        guard let anchor = calendar.dateInterval(of: ruler.unit, for: window.start)?.start
        else { return [] }

        let formatter = DateFormatter()
        formatter.calendar = calendar
        formatter.locale = locale
        formatter.timeZone = calendar.timeZone
        formatter.setLocalizedDateFormatFromTemplate(ruler.unit == .month ? "MMM" : "Md")

        var ticks: [TimelineTick] = []
        var cursor = anchor
        // A pathological window (a decade at day resolution) must not spin here; the ruler is only
        // ever asked for one screen's worth.
        while cursor <= window.end, ticks.count < 400 {
            if cursor >= window.start {
                let major = calendar.dateInterval(of: ruler.major, for: cursor)?.start == cursor
                ticks.append(
                    TimelineTick(date: cursor, label: formatter.string(from: cursor), major: major)
                )
            }
            guard let next = calendar.date(byAdding: ruler.unit, value: ruler.step, to: cursor),
                  next > cursor else { break }
            cursor = next
        }
        return ticks
    }

    /// Groups the sessions that overlap the window, newest group first, splitting each group into
    /// lanes by the dimension the grouping did not take. Sessions inside a lane keep start order so
    /// the lane reads left to right the way the work happened.
    static func groups(
        from sessions: [HistorySessionMetadata],
        grouping: TimelineGrouping,
        window: TimelineWindow
    ) -> [TimelineGroup] {
        let visible = sessions.filter { session in
            let end = max(session.lastActivity, session.createdAt)
            return end >= window.start && session.createdAt <= window.end
        }
        guard !visible.isEmpty else { return [] }

        var order: [String] = []
        var buckets: [String: [HistorySessionMetadata]] = [:]
        for session in visible {
            let key = groupKey(for: session, grouping: grouping)
            if buckets[key] == nil { order.append(key) }
            buckets[key, default: []].append(session)
        }

        return order.compactMap { key -> TimelineGroup? in
            guard let members = buckets[key], let first = members.first else { return nil }
            let lanes = self.lanes(in: members, grouping: grouping)
            switch grouping {
            case .directory:
                return TimelineGroup(
                    id: key,
                    title: directoryTitle(for: first, key: key),
                    subtitle: first.cwd,
                    source: nil,
                    lanes: lanes
                )
            case .agent:
                return TimelineGroup(
                    id: key,
                    title: ConversationPresentation.sourceName(rawValue: first.source.rawValue),
                    subtitle: nil,
                    source: first.source,
                    lanes: lanes
                )
            }
        }
        .sorted { left, right in
            if left.lastActivity != right.lastActivity { return left.lastActivity > right.lastActivity }
            return left.title < right.title
        }
    }

    private static func lanes(
        in members: [HistorySessionMetadata],
        grouping: TimelineGrouping
    ) -> [TimelineLane] {
        var order: [String] = []
        var buckets: [String: [HistorySessionMetadata]] = [:]
        for session in members {
            let key = laneKey(for: session, grouping: grouping)
            if buckets[key] == nil { order.append(key) }
            buckets[key, default: []].append(session)
        }

        return order.compactMap { key -> TimelineLane? in
            guard let members = buckets[key], let first = members.first else { return nil }
            let entries = members
                .sorted {
                    if $0.createdAt != $1.createdAt { return $0.createdAt < $1.createdAt }
                    return $0.id < $1.id
                }
                .map(TimelineEntry.init)
            switch grouping {
            case .directory:
                return TimelineLane(
                    id: key,
                    title: ConversationPresentation.sourceName(rawValue: first.source.rawValue),
                    source: first.source,
                    entries: entries
                )
            case .agent:
                return TimelineLane(
                    id: key,
                    title: directoryTitle(for: first, key: key),
                    source: nil,
                    entries: entries
                )
            }
        }
        .sorted { left, right in
            if left.lastActivity != right.lastActivity { return left.lastActivity > right.lastActivity }
            return left.title < right.title
        }
    }

    private static func directoryTitle(for session: HistorySessionMetadata, key: String) -> String {
        if !session.project.isEmpty { return session.project }
        let leaf = (key as NSString).lastPathComponent
        return leaf.isEmpty ? key : leaf
    }

    private static func groupKey(
        for session: HistorySessionMetadata,
        grouping: TimelineGrouping
    ) -> String {
        grouping == .directory ? directoryKey(for: session) : session.source.rawValue
    }

    private static func laneKey(
        for session: HistorySessionMetadata,
        grouping: TimelineGrouping
    ) -> String {
        grouping == .directory ? session.source.rawValue : directoryKey(for: session)
    }

    private static func directoryKey(for session: HistorySessionMetadata) -> String {
        let cwd = session.cwd ?? ""
        if !cwd.isEmpty { return cwd }
        return session.project.isEmpty ? "—" : session.project
    }
}

File: packages/ui/src/history/timeline-layout.ts
import type { HistoryLocale, HistorySessionSummary, HistorySource } from "./contract.js";

export type TimelineZoom = "week" | "month" | "quarter" | "year";
export type TimelineGrouping = "directory" | "agent";

export interface TimelineWindow {
  start: number;
  end: number;
}

export interface TimelineBar {
  x: number;
  width: number;
}

export interface TimelineTick {
  time: number;
  label: string;
  major: boolean;
}

/** 窄窗口按轨道宽度减少刻度；相邻主刻度优先，避免日期文案互相覆盖。 */
export function selectTimelineTicks(
  ticks: readonly TimelineTick[],
  window: TimelineWindow,
  trackWidth: number,
  minimumSpacing = 56,
): TimelineTick[] {
  if (!Number.isFinite(trackWidth) || trackWidth <= 0 || window.end <= window.start) return [];
  const selected: TimelineTick[] = [];
  const position = (tick: TimelineTick) =>
    ((tick.time - window.start) / (window.end - window.start)) * trackWidth;
  for (const tick of ticks) {
    const x = position(tick);
    // 靠右边缘的标签会被裁成单个字，留出完整标签宽度。
    if (x < 0 || x > trackWidth - minimumSpacing) continue;
    const previous = selected.at(-1);
    if (previous && x - position(previous) < minimumSpacing) {
      if (tick.major && !previous.major) selected[selected.length - 1] = tick;
      continue;
    }
    selected.push(tick);
  }
  return selected;
}

export interface TimelineEntry {
  session: HistorySessionSummary;
  start: number;
  end: number;
}

export interface TimelineLane {
  id: string;
  label: string;
  source: HistorySource | null;
  entries: TimelineEntry[];
  lastActivity: number;
}

export interface TimelineGroup {
  id: string;
  label: string;
  subtitle: string | null;
  source: HistorySource | null;
  lanes: TimelineLane[];
  sessionCount: number;
  lastActivity: number;
}

const DAY = 86_400_000;
const ZOOM_SPANS: Record<TimelineZoom, number> = {
  week: 7 * DAY,
  month: 30 * DAY,
  quarter: 91 * DAY,
  year: 365 * DAY,
};

/** The anchor sits at 92% of the width, leaving a little room after today. */
export function createTimelineWindow(zoom: TimelineZoom, anchor = Date.now()): TimelineWindow {
  const span = ZOOM_SPANS[zoom];
  const end = anchor + span * 0.08;
  return { start: end - span, end };
}

export function shiftTimelineWindow(window: TimelineWindow, fraction: number): TimelineWindow {
  const delta = (window.end - window.start) * fraction;
  return { start: window.start + delta, end: window.end + delta };
}

/** Zoom keeps the visible right edge in place, as in the former CC Buddy calendar. */
export function zoomTimelineWindow(window: TimelineWindow, zoom: TimelineZoom): TimelineWindow {
  return { start: window.end - ZOOM_SPANS[zoom], end: window.end };
}

export function timelineBar(
  start: number,
  end: number,
  window: TimelineWindow,
  trackWidth: number,
  minimumWidth = 6,
): TimelineBar | null {
  if (
    !Number.isFinite(start) ||
    !Number.isFinite(end) ||
    !Number.isFinite(window.start) ||
    !Number.isFinite(window.end) ||
    !Number.isFinite(trackWidth) ||
    window.end <= window.start ||
    trackWidth <= 0 ||
    end < start ||
    end < window.start ||
    start > window.end
  ) {
    return null;
  }

  const span = window.end - window.start;
  const rawStart = ((start - window.start) / span) * trackWidth;
  const rawEnd = ((end - window.start) / span) * trackWidth;
  let x = Math.max(0, rawStart);
  let width = Math.min(trackWidth, rawEnd) - x;
  const minimum = Math.min(Math.max(0, minimumWidth), trackWidth);
  if (width < minimum) {
    width = minimum;
    x = Math.max(0, Math.min(x, trackWidth - minimum));
  }
  return { x, width: Math.min(width, trackWidth - x) };
}

/** Entries are searched in reverse paint order, so a visible overlap opens its top bar. */
export function hitTestTimelineEntry(
  entries: readonly TimelineEntry[],
  window: TimelineWindow,
  trackWidth: number,
  x: number,
  tolerance = 2,
): TimelineEntry | null {
  for (let index = entries.length - 1; index >= 0; index -= 1) {
    const entry = entries[index];
    if (!entry) continue;
    const bar = timelineBar(entry.start, entry.end, window, trackWidth);
    if (bar && x >= bar.x - tolerance && x <= bar.x + bar.width + tolerance) return entry;
  }
  return null;
}

function directoryKey(session: HistorySessionSummary): string {
  return session.cwd?.trim() || session.project.trim() || "—";
}

function directoryLabel(session: HistorySessionSummary): string {
  if (session.project.trim()) return session.project.trim();
  const directory = directoryKey(session);
  return directory.split(/[\\/]/).filter(Boolean).at(-1) || directory;
}

function entryFor(session: HistorySessionSummary): TimelineEntry | null {
  const start = Date.parse(session.createdAt);
  const last = Date.parse(session.lastActivity);
  if (!Number.isFinite(start) || !Number.isFinite(last)) return null;
  return { session, start, end: Math.max(start, last) };
}

/** One source of grouping truth for both rendering and calendar accessibility. */
export function buildTimelineGroups(
  sessions: readonly HistorySessionSummary[],
  grouping: TimelineGrouping,
  window: TimelineWindow,
): TimelineGroup[] {
  const buckets = new Map<string, TimelineEntry[]>();
  for (const session of sessions) {
    const entry = entryFor(session);
    if (!entry || entry.end < window.start || entry.start > window.end) continue;
    const key = grouping === "directory" ? directoryKey(session) : session.source;
    const bucket = buckets.get(key) ?? [];
    bucket.push(entry);
    buckets.set(key, bucket);
  }

  return [...buckets]
    .map(([id, members]): TimelineGroup => {
      const first = members[0]!;
      const laneBuckets = new Map<string, TimelineEntry[]>();
      for (const entry of members) {
        const key = grouping === "directory" ? entry.session.source : directoryKey(entry.session);
        const bucket = laneBuckets.get(key) ?? [];
        bucket.push(entry);
        laneBuckets.set(key, bucket);
      }
      const lanes = [...laneBuckets]
        .map(([laneID, laneEntries]): TimelineLane => {
          laneEntries.sort(
            (left, right) =>
              left.start - right.start || left.session.id.localeCompare(right.session.id),
          );
          return {
            id: laneID,
            label: grouping === "directory" ? laneID : directoryLabel(laneEntries[0]!.session),
            source: grouping === "directory" ? laneEntries[0]!.session.source : null,
            entries: laneEntries,
            lastActivity: laneEntries.reduce(
              (latest, entry) => Math.max(latest, entry.end),
              -Infinity,
            ),
          };
        })
        .sort(
          (left, right) =>
            right.lastActivity - left.lastActivity || left.label.localeCompare(right.label),
        );

      return {
        id,
        label: grouping === "directory" ? directoryLabel(first.session) : first.session.source,
        subtitle: grouping === "directory" ? first.session.cwd : null,
        source: grouping === "agent" ? first.session.source : null,
        lanes,
        sessionCount: members.length,
        lastActivity: members.reduce((latest, entry) => Math.max(latest, entry.end), -Infinity),
      };
    })
    .sort(
      (left, right) =>
        right.lastActivity - left.lastActivity || left.label.localeCompare(right.label),
    );
}

/** Calendar ticks are bounded even when a caller gives an absurd range. */
export function timelineTicks(
  window: TimelineWindow,
  zoom: TimelineZoom,
  locale: HistoryLocale = "zh-CN",
  timeZone?: string,
): TimelineTick[] {
  if (!Number.isFinite(window.start) || !Number.isFinite(window.end) || window.end <= window.start)
    return [];
  const utc = timeZone === "UTC";
  const date = new Date(window.start);
  if (zoom === "year") {
    if (utc) {
      date.setUTCDate(1);
      date.setUTCHours(0, 0, 0, 0);
    } else {
      date.setDate(1);
      date.setHours(0, 0, 0, 0);
    }
  } else {
    if (utc) date.setUTCHours(0, 0, 0, 0);
    else date.setHours(0, 0, 0, 0);
    if (zoom === "quarter") {
      const day = utc ? date.getUTCDay() : date.getDay();
      if (utc) date.setUTCDate(date.getUTCDate() - ((day + 6) % 7));
      else date.setDate(date.getDate() - ((day + 6) % 7));
    }
  }

  const formatter = new Intl.DateTimeFormat(locale, {
    month: "short",
    ...(zoom === "year" ? {} : { day: "numeric" }),
    ...(timeZone ? { timeZone } : {}),
  });
  const ticks: TimelineTick[] = [];
  for (let index = 0; index < 400 && date.getTime() <= window.end; index += 1) {
    const time = date.getTime();
    const day = utc ? date.getUTCDate() : date.getDate();
    const month = utc ? date.getUTCMonth() : date.getMonth();
    const weekday = utc ? date.getUTCDay() : date.getDay();
    if (time >= window.start) {
      ticks.push({
        time,
        label: formatter.format(date),
        major:
          zoom === "year"
            ? month % 3 === 0
            : zoom === "quarter"
              ? day <= 7
              : zoom === "month"
                ? day === 1
                : weekday === 1,
      });
    }
    if (zoom === "year") {
      if (utc) date.setUTCMonth(date.getUTCMonth() + 1);
      else date.setMonth(date.getMonth() + 1);
    } else {
      const days = zoom === "quarter" ? 7 : zoom === "month" ? 3 : 1;
      if (utc) date.setUTCDate(date.getUTCDate() + days);
      else date.setDate(date.getDate() + days);
    }
  }
  return ticks;
}
